File: my-folder/0152-maximum-product-subarray/solution.java
class Solution {
    public int maxProduct(int[] nums) {
        int preMax = nums[0];
        int preMin = nums[0];
        int max = preMax;
        int end = 0;
        for(int i=1;i<nums.length;i++){
            int tempMax = Math.max(nums[i], Math.max(nums[i]*preMax, nums[i]*preMin));
            preMin = Math.min(nums[i], Math.min(nums[i]*preMax, nums[i]*preMin));
            preMax = tempMax;
            
            if(preMax>max){
                max = preMax;
                end = i;
            }
        }
        // int product = 1;
        // int start = end;
        // for(int i=end;i>=0;i--){
        //     product *=nums[i];
        //     if(product==max){
        //         start = i;
        //         break;
        //     }
        // }
        // for(int i=start;i<=end;i++){
        //     System.out.println(nums[i]);
        // }
        
        return max;
    }
}

File: my-folder/0152-maximum-product-subarray/solution.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        curMax, curMin = 1, 1
        finalMax = -10
        for n in nums:
            curMax, curMin = max(n, curMax*n, curMin*n), min(n, curMax*n, curMin*n)
            finalMax = max(curMax, finalMax)
        return finalMax
        
